### ISSUE TITLE
nitropy on Windows does not work

### ISSUE THREAD

**Issue by hase77** [Nitrokey/pynitrokey#272]
Tried my just received nitrokey3 with nitropy version 0.4.29 with Windows11 22H2:

<details>

<summary>Logs </summary>

```
nitropy-0.4.29.exe nk3 version
Command line tool to interact with Nitrokey devices 0.4.29
--- Logging error ---
Traceback (most recent call last):
  File "logging\__init__.py", line 1086, in emit
  File "encodings\cp1252.py", line 19, in encode
UnicodeEncodeError: 'charmap' codec can't encode characters in position 70-72: character maps to <undefined>
Call stack:
  File "nitropy.py", line 15, in <module>
  File "pynitrokey\cli\__init__.py", line 111, in main
  File "click\core.py", line 1130, in __call__
  File "click\core.py", line 1055, in main
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1404, in invoke
  File "click\core.py", line 760, in invoke
  File "click\decorators.py", line 38, in new_func
  File "pynitrokey\cli\nk3\__init__.py", line 451, in version
  File "pynitrokey\cli\nk3\__init__.py", line 81, in connect_device
  File "pynitrokey\cli\nk3\__init__.py", line 60, in list
  File "pynitrokey\nk3\__init__.py", line 24, in list
  File "pynitrokey\nk3\bootloader\__init__.py", line 67, in list
  File "pynitrokey\nk3\bootloader\lpc55.py", line 125, in list
  File "spsdk\mboot\interfaces\usb.py", line 260, in enumerate
  File "libusbsio\libusbsio.py", line 1314, in HIDAPI_Enumerate
  File "logging\__init__.py", line 1434, in debug
  File "logging\__init__.py", line 1589, in _log
  File "logging\__init__.py", line 1599, in handle
  File "logging\__init__.py", line 1661, in callHandlers
  File "logging\__init__.py", line 952, in handle
  File "logging\__init__.py", line 1187, in emit
  File "logging\__init__.py", line 1091, in emit
Message: 'HID enumeration[2067217260672]: device #0: 䑀儘ǡ'
Arguments: ()
Critical error:
No Nitrokey 3 device found
Warning: no backend was found to use for communication. Please refer to documentation how to install additional libraries.
--- Logging error ---
Traceback (most recent call last):
  File "pynitrokey\cli\__init__.py", line 111, in main
  File "click\core.py", line 1130, in __call__
  File "click\core.py", line 1055, in main
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1404, in invoke
  File "click\core.py", line 760, in invoke
  File "click\decorators.py", line 38, in new_func
  File "pynitrokey\cli\nk3\__init__.py", line 451, in version
  File "pynitrokey\cli\nk3\__init__.py", line 83, in connect_device
  File "pynitrokey\cli\nk3\__init__.py", line 67, in _select_unique
pynitrokey.cli.exceptions.CliException: No Nitrokey 3 device found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "logging\__init__.py", line 1086, in emit
  File "encodings\cp1252.py", line 19, in encode
UnicodeEncodeError: 'charmap' codec can't encode characters in position 70-75: character maps to <undefined>
Call stack:
  File "nitropy.py", line 15, in <module>
  File "pynitrokey\cli\__init__.py", line 115, in main
  File "pynitrokey\cli\exceptions.py", line 25, in show
  File "pynitrokey\helpers.py", line 208, in local_critical
  File "pynitrokey\cli\__init__.py", line 95, in list
  File "pynitrokey\cli\__init__.py", line 88, in _list
  File "pynitrokey\cli\nk3\__init__.py", line 118, in list
  File "pynitrokey\nk3\__init__.py", line 24, in list
  File "pynitrokey\nk3\bootloader\__init__.py", line 67, in list
  File "pynitrokey\nk3\bootloader\lpc55.py", line 125, in list
  File "spsdk\mboot\interfaces\usb.py", line 260, in enumerate
  File "libusbsio\libusbsio.py", line 1314, in HIDAPI_Enumerate
  File "logging\__init__.py", line 1434, in debug
  File "logging\__init__.py", line 1589, in _log
  File "logging\__init__.py", line 1599, in handle
  File "logging\__init__.py", line 1661, in callHandlers
  File "logging\__init__.py", line 952, in handle
  File "logging\__init__.py", line 1187, in emit
  File "logging\__init__.py", line 1091, in emit
Message: 'HID enumeration[2067215598912]: device #0: 㘀쑥\uffff\uffff暨丞,'
Arguments: ()
```

```
1384       INFO  libusbsio Loading SIO library: C:\Users\ms\AppData\Local\Temp\_MEI216162\libusbsio\bin\x64\libusbsio.dll
1494       INFO  libusbsio HID enumeration[2067217260672]: initialized
1496      DEBUG  libusbsio HID enumeration[2067217260672]: device #1: CHERRY Keyboard
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #2: CHERRY Keyboard
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #3: CHERRY Keyboard
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #4: CHERRY Keyboard
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #5: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #6: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #7: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #8: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #9: CHERRY Keyboard
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #10: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #11: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #12: CHERRY Keyboard
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #13: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #14: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #15: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #16: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #17: USB Receiver
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #18: AURA LED Controller
1497      DEBUG  libusbsio HID enumeration[2067217260672]: device #19: CHERRY Keyboard
1497       INFO  libusbsio HID enumeration[2067217260672]: finished, total 20 devices
1498      DEBUG pynitrokey.nk3.bootloader.nrf52 Skipping device 20a0:42b2 with ports ['COM3']
1499      DEBUG fido2.hid.windows Failed reading HID descriptor for b'\\\\?\\hid#vid_20a0&pid_42b2&mi_01#9&3290b382&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'
Traceback (most recent call last):
  File "fido2\hid\windows.py", line 390, in list_descriptors
  File "fido2\hid\windows.py", line 280, in get_descriptor
PermissionError: [WinError 5] Zugriff verweigert
1500      DEBUG       root print: Critical error:
1500      DEBUG       root print: No Nitrokey 3 device found
1500      DEBUG       root listing all connected devices:
1500      DEBUG       root :: 'Nitrokey FIDO2' keys
1500      DEBUG       root :: 'Nitrokey Start' keys:
1521      DEBUG       root :: 'Nitrokey 3' keys
1607       INFO  libusbsio HID enumeration[2067215598912]: initialized
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #1: CHERRY Keyboard
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #2: CHERRY Keyboard
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #3: CHERRY Keyboard
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #4: CHERRY Keyboard
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #5: USB Receiver
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #6: USB Receiver
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #7: USB Receiver
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #8: USB Receiver
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #9: CHERRY Keyboard
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #10: USB Receiver
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #11: USB Receiver
1610      DEBUG  libusbsio HID enumeration[2067215598912]: device #12: CHERRY Keyboard
1611      DEBUG  libusbsio HID enumeration[2067215598912]: device #13: USB Receiver
1611      DEBUG  libusbsio HID enumeration[2067215598912]: device #14: USB Receiver
1611      DEBUG  libusbsio HID enumeration[2067215598912]: device #15: USB Receiver
1611      DEBUG  libusbsio HID enumeration[2067215598912]: device #16: USB Receiver
1611      DEBUG  libusbsio HID enumeration[2067215598912]: device #17: USB Receiver
1611      DEBUG  libusbsio HID enumeration[2067215598912]: device #18: AURA LED Controller
1611      DEBUG  libusbsio HID enumeration[2067215598912]: device #19: CHERRY Keyboard
1611       INFO  libusbsio HID enumeration[2067215598912]: finished, total 20 devices
1612      DEBUG pynitrokey.nk3.bootloader.nrf52 Skipping device 20a0:42b2 with ports ['COM3']
1612      DEBUG       root print: --------------------------------------------------------------------------------
1612      DEBUG       root print: Critical error occurred, exiting now
1612      DEBUG       root print: Unexpected? Is this a bug? Would you like to get support/help?
1613      DEBUG       root print: - You can report issues at: https://support.nitrokey.com/
1613      DEBUG       root print: - Writing an e-mail to support@nitrokey.com is also possible
1613      DEBUG       root print: - Please attach the log: 'C:\Users\ms\AppData\Local\Temp\nitropy.log.6zfeb1wz' with any support/help request!
1613      DEBUG       root print: --------------------------------------------------------------------------------
```

</details>

No documentation found what additional libraries should be installed.

**Comment by hase77** [comment]
I also tried the non-exe python version, as admin and non-admin user, but with the same result.

**Comment by robin-nitrokey** [comment]
@hase77 Does this only happen if a Nitrokey 3 device is connected or even without one?  Does `nitropy nk3 list` work?

**Comment by hase77** [comment]
No difference if the key is connected or not. nitropy nk3 list has the same error, as all nk3 other commands like update, test etc..

And I tried to remove almost all other usb devices if one of them cause the problem. Even another keyboard. But nope.

**Comment by Niklap97** [comment]
@hase77, @robin-nitrokey 
I get a similar error and log when i run `nitropy-0.4.29.exe` without admin privileges.
Running the shell as admin fixes it for me.

**Comment by hase77** [comment]
@Niklap97 
Strange. I tried both, as normal user and as adminstrator. No difference. And just in case I tried Windows Terminal and cmd.exe to see if there is a difference.

**Comment by hase77** [comment]
Really weird. I just noticed running as normal user or adminstrator makes a difference. As admin user the version, test or update functions are "working", but the error message stays.

```
PS C:\Users\ms\Desktop> .\nitropy-0.4.30.exe nk3 test --pin XXXX
Command line tool to interact with Nitrokey devices 0.4.30
--- Logging error ---
Traceback (most recent call last):
  File "logging\__init__.py", line 1086, in emit
  File "encodings\cp1252.py", line 19, in encode
UnicodeEncodeError: 'charmap' codec can't encode characters in position 70-72: character maps to <undefined>
Call stack:
  File "nitropy.py", line 15, in <module>
  File "pynitrokey\cli\__init__.py", line 111, in main
  File "click\core.py", line 1130, in __call__
  File "click\core.py", line 1055, in main
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1404, in invoke
  File "click\core.py", line 760, in invoke
  File "click\decorators.py", line 38, in new_func
  File "pynitrokey\cli\nk3\__init__.py", line 266, in test
  File "pynitrokey\cli\nk3\__init__.py", line 60, in list
  File "pynitrokey\nk3\__init__.py", line 24, in list
  File "pynitrokey\nk3\bootloader\__init__.py", line 67, in list
  File "pynitrokey\nk3\bootloader\lpc55.py", line 125, in list
  File "spsdk\mboot\interfaces\usb.py", line 260, in enumerate
  File "libusbsio\libusbsio.py", line 1314, in HIDAPI_Enumerate
  File "logging\__init__.py", line 1434, in debug
  File "logging\__init__.py", line 1589, in _log
  File "logging\__init__.py", line 1599, in handle
  File "logging\__init__.py", line 1661, in callHandlers
  File "logging\__init__.py", line 952, in handle
  File "logging\__init__.py", line 1187, in emit
  File "logging\__init__.py", line 1091, in emit
Message: 'HID enumeration[1482730697424]: device #0: 㑀㪻ř'
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "logging\__init__.py", line 1086, in emit
  File "encodings\cp1252.py", line 19, in encode
UnicodeEncodeError: 'charmap' codec can't encode characters in position 70-72: character maps to <undefined>
Call stack:
  File "nitropy.py", line 15, in <module>
  File "pynitrokey\cli\__init__.py", line 111, in main
  File "click\core.py", line 1130, in __call__
  File "click\core.py", line 1055, in main
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1657, in invoke
  File "click\core.py", line 1404, in invoke
  File "click\core.py", line 760, in invoke
  File "click\decorators.py", line 38, in new_func
  File "pynitrokey\cli\nk3\__init__.py", line 266, in test
  File "pynitrokey\cli\nk3\__init__.py", line 60, in list
  File "pynitrokey\nk3\__init__.py", line 24, in list
  File "pynitrokey\nk3\bootloader\__init__.py", line 67, in list
  File "pynitrokey\nk3\bootloader\lpc55.py", line 125, in list
  File "spsdk\mboot\interfaces\usb.py", line 260, in enumerate
  File "libusbsio\libusbsio.py", line 1314, in HIDAPI_Enumerate
  File "logging\__init__.py", line 1434, in debug
  File "logging\__init__.py", line 1589, in _log
  File "logging\__init__.py", line 1599, in handle
  File "logging\__init__.py", line 1661, in callHandlers
  File "logging\__init__.py", line 952, in handle
  File "logging\__init__.py", line 1187, in emit
  File "logging\__init__.py", line 1091, in emit
Message: 'HID enumeration[1482730697424]: device #1: 㑀㪻ř'
Arguments: ()
Found 1 Nitrokey 3 device(s):
- Nitrokey 3 at \\?\hid#vid_20a0&pid_42b2&mi_01#9&3290b382&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}

Running tests for Nitrokey 3 at \\?\hid#vid_20a0&pid_42b2&mi_01#9&3290b382&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}

[1/3]   uuid            UUID query                      SUCCESS         8FE8BFC45160785B9CFE9B31F837F4FF
[2/3]   version         Firmware version query          SUCCESS         v1.2.2
Please press the touch button on the device ...
Please press the touch button on the device ...
[3/3]   fido2           FIDO2                           SUCCESS

3 tests, 3 successful, 0 skipped, 0 failed

Summary: 1 device(s) tested, 1 successful, 0 failed
```

**Comment by szszszsz** [comment]
It's a logging error, coming from the cp1252 charmap. Nothing important in this case, but should be fixed anyway. Why utf-8 is not used in the first place?

**Comment by robin-nitrokey** [comment]
So we are seeing two different errors here:
- The one with `UnicodeEncodeError: 'charmap' codec can't encode characters in position 70-72: character maps to <undefined>` always occurs and is caused by the log message in `libusbsio`.  It clutters the output but it is not critical.  I will report it to upstream via mail.
- But there also is this snippet which indicates a permissions error.  This one seems to be fixed by running `nitropy` with admin privileges.

```
1499      DEBUG fido2.hid.windows Failed reading HID descriptor for b'\\\\?\\hid#vid_20a0&pid_42b2&mi_01#9&3290b382&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'
Traceback (most recent call last):
  File "fido2\hid\windows.py", line 390, in list_descriptors
  File "fido2\hid\windows.py", line 280, in get_descriptor
PermissionError: [WinError 5] Zugriff verweigert
```

**Comment by hase77** [comment]
Ok, then nitrpy with admin rights works on windows. But with an misleading message. Maybe nitropy can check for runnning as administrator on windows?

**Comment by robin-nitrokey** [comment]
I’m not sure whether execution as a regular user always fails or if it depends on the configuration.  @szszszsz Do you know more about that?  And unfortunately, we cannot detect this error during device enumeration because it is discarded by the `fido2` library.

I’ve tried to fix the encoding error by enforcing UTF-8 for the log file in [this patch](https://github.com/Nitrokey/pynitrokey/pull/276).  It looks like Python defaults to using `cp1252` when opening files on Windows.

**Comment by szszszsz** [comment]
That depends on the device, but accessing FIDO2 devices (and all its subdevices shall it have any) directly is excluded for non-Admin user since Windows 10 (instead the WinAPI usage is encourage).

AFAIR detecting whether user has Admin privileges should be easy. @robin-nitrokey can you add that as well?

**Comment by robin-nitrokey** [comment]
@szszszsz For which subcommands would you add the check?  At least `nethsm` does not need admin privileges.

**Comment by szszszsz** [comment]
All concerning FIDO2 devices on Windows are a must (`fido2` and `nk3`). I do not think this is needed for other.

**Comment by hase77** [comment]
Found by coincidence the usb device which is causing the misleading error message. It's the virtual usb-c connector from the nvidia 2060 gfx card. I switch to my old nvidia 970 without "usb-controller" and no error message.

**Comment by szszszsz** [comment]
That's weird indeed. Either way, this should be fixed in the 0.4.31 release.
